test/dynamo/test_subclasses.py

```diff
@@ -128,6 +129,119 @@ def f(x):
         self.assertEqual(cnt.frame_count, 3)
         self.assertEqual(cnt.op_count, 3)
 
+    def test_wrapper_subclass_guards_on_inner_tensor(self):
+        # Holds an inner tensor, that has a distinct shape from the outer wrapper tensor.
+        # Also adds additional guards on the inner tensor's sizes.
+        # When the first input to an op has x.shape[0] > 5, we insert an extra add node.
+        class DoubleSizeMaybeAddGeThreeTensor(torch.Tensor):
+            @staticmethod
+            def __new__(cls, inner):
+                # Double the outer-most dimension
+                outer_shape = (inner.shape[0] * 2,) + inner.shape[1:]
+                return torch.Tensor._make_wrapper_subclass(
+                    # TODO: right now, _make_wrapper_subclass's dynamic shape interaction is not great.
+                    # Calling the overload that has kwargs causes us to go down the first overload path,
+                    # which will **always** specialize sizes.
+                    # We should probably eventually fix this so that the first overload can just handle dynamic shapes.
+                    cls,
+                    outer_shape,
+                    inner.stride(),
+                    None,
+                    None,
+                    inner.dtype,
+                    inner.layout,
+                    inner.device,
+                    False,
+                    inner.requires_grad,
+                )
+
+            def __init__(self, inner):
+                self.inner_elem = inner
+
+            def __tensor_flatten__(self):
+                return {"inner_elem": self.inner_elem}, None
```

Comment:
Huh, I didn't realize we were using a dict for tensor flatten. Are you sure? Doing hash string lookups is going to be less efficient than list index lookups

Reply:
Oh, is the point that the key has to exactly match the attribute name? Then why do we need to return the Tensors themselves?

Reply:
Going with the updated version of approach (1) further down